### PR TITLE
KAFKA-17796: Persist higher leaderEpoch in read state call.

### PR DESCRIPTION
**Pull request by smjn**
It is possible that due to a network partition or delays on the wire a read request with a higher `leaderEpoch` reaches the share coordinator before a write request with a lower one for the same share partition. In this case, we have chosen to honor the highest `leaderEpoch` request and reject any subsequent lower value requests irrespective of the RPC type. However, this means that we need to persist the `leaderEpoch` seen in the read requests.

In the `ShareCoordinatorShard.readState` method we are storing the `leaderEpoch` present in the request directly in the `leaderEpochMap` timeline hashmap, if it is the highest `leaderEpoch` seen so far for a specific share partition. This however is folly. 

The coordinator runtime guarantees the consistency of the timeline data structures when updated via the `replay` method. The `replay` method is called by the runtime whenever it persists some records into the topic it is managing. This method is implemented in `ShareCoordinatorShard.replay`.

Therefore, to remedy the situation this PR adds code into the `ShareCoordinatorService.readState` method to issue a `runtime.scheduleWriteOperation` call if the incoming read state request holds a valid `leaderEpoch` value (not -1). We cannot ascertain if the `leaderEpoch` in the read request is the highest so far for the given share partition because it is not possible to lookup the timeline data structures in the `ShareCoordinatorService` class since they are housed in `ShareCoordinatorShard` and an offset is needed to do the lookup, which we won't have. Secondly, we cannot move code to write the record in the shard read state method as we need the runtime to execute that callback which is present in the `ShareCoordinatorService`.

Hence, the new code will issue a `scheduleWriteOperation` call if leaderEpoch is != -1 and the supplied callback method `ShareCoordinatorShard.writeLeaderEpoch` (also part of the PR) will generate a record if the `leaderEpoch` is highest, or no record if it is same as last seen, or error if the epoch is old. We required a separate method because we want to only look at `leaderEpoch` and simply ignore other data fields. Also, we want to perform the optimization of not generating records if the `leaderEpoch` is -1 or equal to highest seen so far.

Subsequently, a sequential call to the `runtime.scheduleReadOperation` will be made in `ShareCoordinatorService.readState` which is same as before.

**TLDR**: We will issue a phantom write call (not explicitly issued by a caller) in `ShareCoordinatorService.readState` if the read request contains a valid `leaderEpoch`. Based on the response to the write, a standard read call will be scheduled. This way we will be able to persist the `leaderEpoch` correctly in the topic and timeline data structures.

### CODE REVIEW

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java [hunk 1]
```diff
@@ -289,6 +289,56 @@ public CoordinatorResult<WriteShareGroupStateResponseData, CoordinatorRecord> wr
         return new CoordinatorResult<>(Collections.singletonList(record), responseData);
     }
 
+    /**
+     * This is a special case method when only the leaderEpoch needs to be updated.
+     * It can happen if a read state call for a share partition has the highest leaderEpoch
```
Comment (AndrewJSchofield) [inline share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java]: It can happen if a read state call for a share partition has a higher leaderEpoch value than seen so far.

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java [hunk 2]
```diff
@@ -478,6 +495,74 @@ public CompletableFuture<ReadShareGroupStateResponseData> readState(RequestConte
...
+                    .setPartitions(Collections.singletonList(new WriteShareGroupStateRequestData.PartitionData()
+                        .setPartition(partitionData.partition())
+                        .setLeaderEpoch(partitionData.leaderEpoch())
+                        // force below attributes to be noop
+                        .setStateEpoch(-1)
+                        .setStateBatches(Collections.emptyList())
+                        .setStartOffset(-1))))
+                );
+
+
```
Comment (AndrewJSchofield) [inline share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java]: nit: Lots of blank lines here.

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java [hunk 3]
```diff
@@ -289,6 +289,56 @@ public CoordinatorResult<WriteShareGroupStateResponseData, CoordinatorRecord> wr
...
+        Optional<CoordinatorResult<WriteShareGroupStateResponseData, CoordinatorRecord>> error = maybeGetWriteStateError(request);
+        if (error.isPresent()) {
+            return error.get();
+        }
+
+        WriteShareGroupStateRequestData.WriteStateData topicData = request.topics().get(0);
+        WriteShareGroupStateRequestData.PartitionData partitionData = topicData.partitions().get(0);
+        SharePartitionKey key = SharePartitionKey.getInstance(request.groupId(), topicData.topicId(), partitionData.partition());
+
+        int leaderEpoch = request.topics().get(0).partitions().get(0).leaderEpoch();
```
Comment (AndrewJSchofield) [inline share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java]: This can be `partitionData.leaderEpoch()` to save some lookups.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
Thanks for the PR. The overall implementation looks good to me.

**Review by AndrewJSchofield — APPROVED** [review]
lgtm. Thanks for the PR.

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java [hunk 4]
```diff
@@ -419,37 +419,54 @@ public CompletableFuture<ReadShareGroupStateResponseData> readState(RequestConte
...
+                    ReadShareGroupStateRequestData requestForCurrentPartition = new ReadShareGroupStateRequestData()
+                        .setGroupId(groupId)
+                        .setTopics(Collections.singletonList(new ReadShareGroupStateRequestData.ReadStateData()
+                            .setTopicId(topicId)
+                            .setPartitions(Collections.singletonList(partitionData))));
+                    // Scheduling a runtime read operation to read share partition state from the coordinator in memory state
+                    return runtime.scheduleReadOperation(
+                        "read-share-group-state",
+                        topicPartitionFor(coordinatorKey),
+                        (coordinator, offset) -> coordinator.readState(requestForCurrentPartition, offset)
```
Comment (dajac) [inline share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java]: It is a slightly orthogonal question but is it safe to read the state at the last committed offset here? This means that you can potentially go back to the state at the committed offset after a share leader failover instead of starting from the last published state. Think about the following:
* share leader writes new state at offset 20. this is is not committed yet.
* share leader fails over to another replica. it reads the state based on the last committed offset, say 10 in this case.
* offset 20 is committed afterwards.

I am not sure whether this is a realistic scenario or whether they are other cases where we write and the read. If we want a strong read after write consistency, it may be better to use a write operation to read too because it guarantees the ordering. If you do something like this, you could also combine updating the leader epoch in the same operation.

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java [hunk 5]
```diff
@@ -296,6 +296,94 @@ public CoordinatorResult<WriteShareGroupStateResponseData, CoordinatorRecord> wr
...
+        if (!shareStateMap.containsKey(key)) {
+            responseData = ReadShareGroupStateResponse.toResponseData(
+                topicId,
+                partitionId,
+                PartitionFactory.UNINITIALIZED_START_OFFSET,
+                PartitionFactory.DEFAULT_STATE_EPOCH,
+                Collections.emptyList()
+            );
+        } else if (shareStateMap.get(key) == null) {
+            // no state information was not found but key present
```
Comment (AndrewJSchofield) [inline share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java]: nit: double negative. I think you mean that no share state information was found but key present

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java [hunk 6]
```diff
@@ -296,6 +296,94 @@ public CoordinatorResult<WriteShareGroupStateResponseData, CoordinatorRecord> wr
...
 
+    /**
+     * Method reads data from the soft state and if needed updates the leader epoch.
+     * It can happen that a read state call for a share partition has a higher leaderEpoch
+     * value than seen so far.
+     * In case an update is not required, empty record list will be generated along with a success response.
+     * @param request - represents ReadShareGroupStateRequestData
+     * @return CoordinatorResult object
+     */
+    public CoordinatorResult<ReadShareGroupStateResponseData, CoordinatorRecord> maybeUpdateLeaderEpochAndRead(
```
Comment (AndrewJSchofield) [inline share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java]: Personally, I think this is `readStateAndMaybeUpdateLeaderEpoch`. It's the read method, that is also able to update the leader epoch if necessary.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
Thanks for the PR. I was bothered about the leader epoch map being slightly soft state in the previous iteration, but this is much more convincing.

**Review by AndrewJSchofield — APPROVED** [review]
lgtm. Thanks for the PR.

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java [hunk 7]
```diff
@@ -296,6 +296,94 @@ public CoordinatorResult<WriteShareGroupStateResponseData, CoordinatorRecord> wr
...
+     * It can happen that a read state call for a share partition has a higher leaderEpoch
+     * value than seen so far.
+     * In case an update is not required, empty record list will be generated along with a success response.
+     * @param request - represents ReadShareGroupStateRequestData
+     * @return CoordinatorResult object
+     */
+    public CoordinatorResult<ReadShareGroupStateResponseData, CoordinatorRecord> readStateAndMaybeUpdateLeaderEpoch(
+        ReadShareGroupStateRequestData request
+    ) {
+        // only one key will be there in the request by design
```
Comment (dajac) [inline share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java]: small nit: it noticed that your comments are inconsistent. sometimes they start with a capital letter, sometimes they don't. sometimes they end with a `.`, sometimes they don't. i don't have a preference for any format but i do care about consistency :)

**Review by dajac — COMMENTED** [review]
lgtm. left a small nit.

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java (hunk 7 above)
Comment (smjn) [reply share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java]: if its a single sentence/phrase i don't pay attention to dots and case

If its a para, the standard english formatting.

Will change all

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java [hunk 8]
```diff
@@ -296,6 +296,102 @@ public CoordinatorResult<WriteShareGroupStateResponseData, CoordinatorRecord> wr
...
+        if (!shareStateMap.containsKey(key)) {
+            // Leader epoch update might be needed
+            responseData = ReadShareGroupStateResponse.toResponseData(
+                topicId,
+                partitionId,
+                PartitionFactory.UNINITIALIZED_START_OFFSET,
+                PartitionFactory.DEFAULT_STATE_EPOCH,
+                Collections.emptyList()
+            );
+        } else if (shareStateMap.get(key) == null) {
```
Comment (dajac) [inline share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java]: I have a question regarding this branch. If I recall correctly, TimelineHashMap does not allow null keys/values. Hence, I am not sure if we really need it. I was looking into it because you get the key three times in a row here on the happy path.

File: build.gradle [hunk 9]
```diff
@@ -203,6 +203,11 @@ allprojects {
     options.links "https://docs.oracle.com/en/java/javase/${JavaVersion.current().majorVersion}/docs/api/"
   }
 
+  tasks.withType(Checkstyle) {
```
Comment (AndrewJSchofield) [inline build.gradle]: I expect this is no longer required since the memory shortage was diagnosed and fixed separately.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
Apart from the question about the change in build.gradle, this PR looks good to me. I've tested it manually on my local system, both with a clean broker and a restarted broker. I can see the leader epoch incrementing by dumping the log.

Happy to merge this once the query about the build.gradle is answered.

File: build.gradle (hunk 9 above)
Comment (smjn) [reply build.gradle]: @AndrewJSchofield synced with trunk - the checkstyle change is no longer part of this PR

**Review by AndrewJSchofield — APPROVED** [review]
